cmd/service/deploy.go

```diff
@@ -24,11 +29,125 @@ To get more information, see the [deploy page from the documentation](https://do
 }
 
 func deployHandler(cmd *cobra.Command, args []string) {
-	service := prepareService(defaultPath(args))
-	reply, err := cli().DeployService(context.Background(), &core.DeployServiceRequest{
-		Service: service,
-	})
+	path := defaultPath(args)
+	serviceID, isValid, err := deployService(path)
+	if !isValid {
+		os.Exit(1)
+	}
 	utils.HandleError(err)
-	fmt.Println("Service deployed with ID:", aurora.Green(reply.ServiceID))
-	fmt.Println("To start it, run the command: mesg-core service start " + reply.ServiceID)
+
+	fmt.Println("Service deployed with ID:", aurora.Green(serviceID))
+	fmt.Printf("To start it, run the command: mesg-core service start %s\n", serviceID)
+
+}
+
+func deployService(path string) (serviceID string, isValid bool, err error) {
+	stream, err := cli().DeployService(context.Background())
+	if err != nil {
+		return "", true, err
+	}
+
+	deployment := make(chan deploymentResult)
+	go readDeployReply(stream, deployment)
+
+	if govalidator.IsURL(path) {
+		if err := stream.Send(&core.DeployServiceRequest{
+			Value: &core.DeployServiceRequest_Url{Url: path},
+		}); err != nil {
+			return "", true, err
+		}
+	} else {
+		if err := deployServiceSendServiceContext(path, stream); err != nil {
+			return "", true, err
+		}
+	}
+
+	if err := stream.CloseSend(); err != nil {
+		return "", true, err
+	}
+
+	result := <-deployment
+	return result.serviceID, result.isValid, result.err
+}
+
+func deployServiceSendServiceContext(path string, stream core.Core_DeployServiceClient) error {
+	archive, err := archive.TarWithOptions(path, &archive.TarOptions{
+		Compression: archive.Gzip,
+	})
+	if err != nil {
+		return err
+	}
+
+	buf := make([]byte, 1024)
+	for {
+		n, err := archive.Read(buf)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if err := stream.Send(&core.DeployServiceRequest{
+			Value: &core.DeployServiceRequest_Chunk{Chunk: buf[:n]},
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type deploymentResult struct {
+	serviceID string
+	err       error
+	isValid   bool
+}
+
+func readDeployReply(stream core.Core_DeployServiceClient, deployment chan deploymentResult) {
+	sp := spinner.New(utils.SpinnerCharset, utils.SpinnerDuration)
```

Comment:
What don't you use `utils.StartSpinner` or `utils.ShowSpinnerForFunc`?

Reply:
If the existing helper function are not useful for what you are doing, I suggest to create a helper function in order to reuse it if it make sense.

Reply:
I had to create it manually in the first for some reason... I'll see if i can use current util for this

Reply:
done